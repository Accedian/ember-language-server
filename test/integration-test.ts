import * as cp from 'child_process';
import * as path from 'path';
import * as fs from 'fs';

import { createMessageConnection, MessageConnection, Logger, IPCMessageReader, IPCMessageWriter } from 'vscode-jsonrpc';
import {
  DidOpenTextDocumentNotification,
  InitializeRequest,
  CompletionRequest
} from 'vscode-languageserver-protocol';

function startServer() {
  const serverPath = './lib/start-server.js';

  return cp.fork(serverPath, [], {
    cwd: path.join(__dirname, '..')
  });
}

function openFile(connection: MessageConnection, filePath: string) {
  connection.sendNotification(DidOpenTextDocumentNotification.type, {
    textDocument: {
      uri: `file://${filePath}`,
      text: fs.readFileSync(filePath, 'utf8')
    }
  });
}

describe('integration', function() {
  let connection: MessageConnection;
  let serverProcess: cp.ChildProcess;

  beforeAll(() => {
    serverProcess = startServer();
    connection = createMessageConnection(
      new IPCMessageReader(serverProcess),
      new IPCMessageWriter(serverProcess),
      <Logger>{
        error(msg) { console.log('error', msg); },
        log(msg) { console.log('log', msg); },
        info(msg) { console.log('info', msg); },
        warn(msg) { console.log('warn', msg); },
      }
    );

    connection.listen();
  });

  afterAll(() => {
    connection.dispose();
    serverProcess.kill();
  });

  describe('Initialize request', () => {
    it('returns an initialize request', async () => {
      const params = {
        rootUri: `file://${path.join(__dirname, 'fixtures', 'full-project')}`,
        capabilities: {}
      };

      const response = await connection.sendRequest(InitializeRequest.type, params);

      expect(response).toMatchSnapshot();
    });
  });

  describe('Completion request', () => {
    it('returns all components and helpers when requesting completion items in a handlebars expression', async () => {
      const applicationTemplatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'application.hbs');
      const params = {
        textDocument: {
          uri: `file://${applicationTemplatePath}`
        },
        position: {
          line: 1,
          character: 2
        }
      };

      openFile(connection, applicationTemplatePath);

      const response = await connection
        .sendRequest(CompletionRequest.type, params);

      expect(response).toMatchSnapshot();
    });
  });
});
