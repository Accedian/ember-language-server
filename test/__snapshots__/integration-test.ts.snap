// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration Completion request returns all components and helpers when requesting completion items in a handlebars expression 1`] = `
Array [
  Object {
    "detail": "component",
    "kind": 7,
    "label": "my-awesome-component",
  },
  Object {
    "kind": 7,
    "label": "another-awesome-component",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "action",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember 1.11.0",
    "kind": 3,
    "label": "component",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.11.0",
  },
  Object {
    "detail": "Ember 1.13.0",
    "kind": 3,
    "label": "concat",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.13.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "debugger",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember 2.1.0",
    "kind": 3,
    "label": "get",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "2.1.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "if",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "input",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "link-to",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "loc",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "log",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "mount",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "outlet",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "partial",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "render",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "textarea",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unbound",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unless",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
]
`;

exports[`integration Initialize request returns an initialize request 1`] = `
Object {
  "capabilities": Object {
    "completionProvider": Object {
      "resolveProvider": true,
    },
    "definitionProvider": true,
    "documentSymbolProvider": true,
    "textDocumentSync": 1,
  },
}
`;
